### PR TITLE
CLN: always dispatch-to-series

### PR DESCRIPTION
**Pull request by jbrockmendel**
A couple more cleanups follow this.

### CODE REVIEW

File: pandas/core/ops/__init__.py [hunk 1]
```diff
@@ -624,7 +629,7 @@ def to_series(right):
 
 
 def _should_reindex_frame_op(
-    left: "DataFrame", right, op, axis, default_axis: int, fill_value, level
+    left: "DataFrame", right, op, axis, default_axis, fill_value, level
```
Comment (simonjayhawkins) [inline pandas/core/ops/__init__.py:632]: why is this changed?

Comment (jbrockmendel) [reply pandas/core/ops/__init__.py:632]: without this i got mypy complaints that i havent been able to track down

Comment (WillAyd) [reply pandas/core/ops/__init__.py:632]: Can you share the error message?

Comment (jbrockmendel) [reply pandas/core/ops/__init__.py:632]: `pandas/core/ops/__init__.py:706: error: Argument 5 to "_should_reindex_frame_op" has incompatible type "Optional[str]"; expected "int"`

Comment (WillAyd) [reply pandas/core/ops/__init__.py:632]: Gotcha. Looks like `_get_frame_op_default_axis` returns an Optional string and that is used as the default argument in the `f` function local to `_arith_method_FRAME`.

So makes sense Mypy would complain about that. Whether the existing annotation is wrong or if its discovering a bug for us I'm not sure

Comment (jbrockmendel) [reply pandas/core/ops/__init__.py:632]: i think its the annotation.  there is some more code that can be ripped out after this, so im planning to do an annotation pass towards the end

File: pandas/core/ops/__init__.py [hunk 2]
```diff
@@ -324,7 +322,10 @@ def dispatch_to_series(left, right, func, str_rep=None, axis=None):
         return type(left)(bm)
 
     elif isinstance(right, ABCDataFrame):
-        assert right._indexed_same(left)
+        assert left.index.equals(right.index)
+        assert left.columns.equals(right.columns)
+        # TODO: The previous assertion `assert right._indexed_same(left)`
+        #  fails in cases with shaoe[1] == 0
```
Comment (jorisvandenbossche) [inline pandas/core/ops/__init__.py:328]: I don't really understand this comment

File: pandas/core/ops/__init__.py [hunk 3]
```diff
@@ -701,18 +707,19 @@ def f(self, other, axis=default_axis, level=None, fill_value=None):
...
-            pass_op = pass_op if not is_logical else op
-
-            new_data = self._combine_frame(other, pass_op, fill_value)
+            new_data = self._combine_frame(other, na_op, fill_value)
 
         elif isinstance(other, ABCSeries):
             # For these values of `axis`, we end up dispatching to Series op,
             # so do not want the masked op.
+            # TODO: the above comment is no longer accurate since we now
+            #  operate blockwise if other._values is an ndarray
```
Comment (jorisvandenbossche) [inline pandas/core/ops/__init__.py:722]: Can you update the comment instead?

File: pandas/core/ops/array_ops.py [hunk 4]
```diff
@@ -386,8 +379,13 @@ def get_array_op(op, str_rep: Optional[str] = None):
 
     Returns
     -------
-    function
+    functools.partial
     """
+    if isinstance(op, partial):
+        # We get here via dispatch_to_series in DataFrame case
+        # TODO: avoid getting here
```
Comment (jorisvandenbossche) [inline pandas/core/ops/array_ops.py:386]: What is an example with which you get this?

**Review by jorisvandenbossche — COMMENTED** [review]
Generally looks good, few small questions/comments

File: pandas/core/ops/array_ops.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/ops/array_ops.py:386]: In _arith_method_FRAME we pass na_op (which is one a partial returned by this function here) to _combine_frame, which in turn passes na_op to dispatch_to_series, which then passes na_op back to this function

File: pandas/core/ops/__init__.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/ops/__init__.py:722]: simpler to do this in the next pass, which will make us always pass the same thing, at which point we can remove the comment entirely

File: pandas/core/ops/__init__.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/ops/__init__.py:328]: im finding cases where left.columns and right.columns are both empty but different dtypes, which still satisfy `left.columns.equals(right.columns)`, but not `right._indexed_same(left)`.  looks like this is an artifact of the intersection done in `_should_reindex_frame_op`, looking into this now.  (also will fix the typo shaoe -> shape)

File: pandas/core/ops/__init__.py [hunk 5]
```diff
@@ -680,11 +685,12 @@ def _frame_arith_method_with_reindex(
 
 
 def _arith_method_FRAME(cls, op, special):
+    # This is the only function where `special` can be either True or False
```
Comment (jreback) [inline pandas/core/ops/__init__.py:688]: can you type if possible

Comment (jbrockmendel) [reply pandas/core/ops/__init__.py:689]: done

**Comment by jreback** [comment]
ahh need to rebase

**Comment by jbrockmendel** [comment]
Looks like the combination of #34324+#34323 have a subtle merge conflict that broke master.  this should be fixed by any of my just-rebased PRs

**Comment by jreback** [comment]
kk ping when ready in one

**Comment by jbrockmendel** [comment]
pseudo-green. the remaining CI fail is anoteher ipython directive thing in the docbuild

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks

if u wouldn’t mind pushing a change for the doc thing
